File: 2022/day_04.rb
#!/usr/bin/env ruby

PATH = ARGV[0] || File.join(__dir__, "inputs", "example_04.txt")

class Range
  def disjoint?(other)
    not self.intersect?(other)
  end

  def intersect?(other)
    # Uses `self.max` to support exclusive ranges correctly.
    return true if self.begin <= other.begin && other.begin <= self.max
    return true if other.begin <= self.begin && self.begin <= other.max
  end

  def intersection(other)
    return nil if self.disjoint?(other)
    Range.new([self.begin, other.begin].max, [self.max, other.max].min)
  end

  alias_method :&, :intersection
end

camp_assignment_pairs = File.read(PATH).chomp.split("\n")
  .map { |section_pair|
  section_pair
    .split(/,|-/)
    .map(&:to_i)
    .each_slice(2)
    .map { Range.new(_1, _2) }
}

def part_1(data)
  data.count { _1.cover?(_2) || _2.cover?(_1) }
end

def part_2(data)
  data.count { _1.intersect?(_2) }
end

puts part_1(camp_assignment_pairs)
puts part_2(camp_assignment_pairs)

File: 2023/day_04.exs
#!/usr/bin/env elixir

Mix.install([
  {:nimble_parsec, "~> 1.0"}
])

defmodule Day04 do
  def parse_card_wins(records) do
    {:ok, parsed, "", _, _, _} = Day04.Parser.from_string(records)

    Enum.map(parsed, fn {:game, game} ->
      {
        game[:id],
        MapSet.new(game[:winning_numbers])
        |> MapSet.intersection(MapSet.new(game[:numbers_we_have]))
        |> MapSet.size()
      }
    end)
  end

  def part1(records) do
    parse_card_wins(records)
    |> Enum.map(&floor(2 ** (elem(&1, 1) - 1)))
    |> Enum.sum()
  end

  def part2(records) do
    card_wins = parse_card_wins(records)
    original_cards = Map.new(card_wins, &{elem(&1, 0), 1})

    Enum.reduce(card_wins, original_cards, fn {id, count}, acc ->
      Range.new(id + 1, id + count, 1)
      |> Enum.reduce(%{}, fn i, iacc ->
        Map.put(iacc, i, Map.get(acc, id, 1))
      end)
      |> Map.merge(acc, fn _, v1, v2 -> v1 + v2 end)
    end)
    |> Map.values()
    |> Enum.sum()
  end
end

defmodule Day04.Parser do
  import NimbleParsec

  number =
    ignore(optional(string(" ")))
    |> integer(min: 1, max: 2)
    |> ignore(optional(string(" ")))

  numbers = fn type ->
    times(number, min: 1)
    |> tag(type)
  end

  game =
    ignore(string("Card"))
    |> ignore(times(string(" "), min: 1))
    |> integer(min: 1)
    |> unwrap_and_tag(:id)
    |> ignore(string(": "))
    |> concat(numbers.(:winning_numbers))
    |> ignore(string("| "))
    |> concat(numbers.(:numbers_we_have))
    |> tag(:game)
    |> ignore(optional(string("\n")))

  games = repeat(game)

  defparsec(:from_string, games)
end

data = File.read!("./inputs/example_04.txt")

IO.inspect(Day04.part1(data))
IO.inspect(Day04.part2(data))
